### PR TITLE
Feature/exporter progress report

### LINKED ISSUE #8308: Report progress on local media creation / progressing

### PR DESCRIPTION
**Pull request by SergioEstevao**
Fixes #8308 

This PR adds progress reporting and the possibility of cancelation during the export of media from the Photo library to the app.

This is specially important when using iCloud media, where the media may need to be downloaded before being used and on Video insertion where the export process can take the same time has the upload.

This was implemented by making all the exporters reporting their progress and hooking this information with the `createMedia` method on `MediaService` and then connecting this to the addMedia method on the `MediaCoordinator`.

To test:
 - Open the Media Library of a Site
 - Tap to add new media
 - Select the Async option
 - Select media that is stored on iCloud or/and that is a large video
 - See if the initial part display progress.

### CODE REVIEW

**Comment by SergioEstevao** [comment]
@frosty do you mind to give a look on this :).

**Comment by SergioEstevao** [comment]
@frosty do you mind to give a look to this?

File: WordPress/Classes/Utility/Media/MediaAssetExporter.swift [hunk 1]
```diff
@@ -50,172 +50,198 @@ class MediaAssetExporter: MediaExporter {
...
+        if exporterVideoOptions.preferredExportVideoType == nil {
+            exporterVideoOptions.preferredExportVideoType = videoResource.uniformTypeIdentifier
+        }
+
+        let originalFilename = videoResource.originalFilename
+
+        // Request an export session, which may take time to download the complete video data.
+        let options = PHVideoRequestOptions()
+        options.isNetworkAccessAllowed = true
+        let progress = Progress.discreteProgress(totalUnitCount: 100)
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaAssetExporter.swift]: See earlier comments for these 100 / 50 values. Perhaps we can even have some helper methods on `Progress` to encapsulate what this is doing.

File: WordPress/Classes/Utility/Media/MediaVideoExporter.swift [hunk 2]
```diff
@@ -48,78 +48,106 @@ class MediaVideoExporter: MediaExporter {
...
-                        onError(self.exporterErrorWith(error: error))
-                    } else {
-                        onError(VideoExportError.failedExportingVideoDuringExportSession)
-                    }
-                    return
+        // Configure metadata filter for sharing, if we need to remove location data.
+        if options.stripsGeoLocationIfNeeded {
+            session.metadataItemFilter = AVMetadataItemFilter.forSharing()
+        }
+        let progress = Progress.discreteProgress(totalUnitCount: 100)
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaVideoExporter.swift]: Same as other `100` comments.

File: WordPress/Classes/Utility/Media/MediaVideoExporter.swift [hunk 3]
```diff
@@ -128,36 +156,41 @@ class MediaVideoExporter: MediaExporter {
...
+        guard asset.isExportable else {
+            onError(exporterErrorWith(error: VideoExportError.videoAssetWasDetectedAsNotExportable))
+            return Progress.discreteCompletedProgress()
+        }
+        let generator = AVAssetImageGenerator(asset: asset)
+        if let imageOptions = imageOptions, let maxSize = imageOptions.maximumImageSize {
+            generator.maximumSize = CGSize(width: maxSize, height: maxSize)
+        }
+        generator.appliesPreferredTrackTransform = true
+        let progress = Progress.discreteProgress(totalUnitCount: 2)
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaVideoExporter.swift]: Why 2 and 1 here?

**Comment by SergioEstevao** [comment]
> Could we just always use 0 to 1.0 for these?

Unfortunately these values need to be integers, but you are right in that I always should try to use the same scale, I will convert them to 0 to 100 all of them, and use constants.

**Comment by SergioEstevao** [comment]
> Thumbnails don't seem to be appearing in the cell as reliably as they were before. Most of the time they're not appearing (navy blue background), sometimes I see the image placeholder icon, and once I've seen an actual thumbnail

Just to clarify, I didn't change the moment when thumbnails are generated, they are still being created **after** the export process is done.

**Comment by SergioEstevao** [comment]
@frosty thanks for your review, I update the code to address your comments can you give it another look.

**Comment by frosty** [comment]
Code changes look good – much more readable with the constants 👍 

However, I'm still seeing the thumbnail issue – here I chose 5 recent images to upload and all but one of them fail to show thumbnails:

img_870becd5bd4c-1

I also saw the following behaviour with a large video from a long way back in my media library - I see indeterminate progress for a long time (I guess while the video is processed), and then when the discrete progress appears it starts at 25%:

upload-progress-2

File: WordPress/Classes/Utility/Media/MediaVideoExporter.swift [hunk 4]
```diff
@@ -208,3 +214,34 @@ class MediaVideoExporter: MediaExporter {
...
+        self.videoSession = videoSession
+        self.progressHandler = progressHandler
+        interrupt = false
+        self.work()
+    }
+
+    private func work(){
+        DispatchQueue.global().asyncAfter(deadline: DispatchTime.now() + DispatchTimeInterval.milliseconds(100)) {
+            self.progressHandler(self.videoSession.progress)
+            if (self.videoSession.progress != 1 && !self.interrupt) {
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaVideoExporter.swift]: Perhaps it'd be safest to do `< 1` here instead of `!= 1`?

**Pull request merged**